flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java

```diff
@@ -120,7 +120,7 @@
 
     private final OperatorEventDispatcherImpl operatorEventDispatcher;
 
-    private boolean isStoppingBySyncSavepoint;
+    private volatile boolean isStoppingBySyncSavepoint;
```

Comment:
This would introduce new "synchronisation" point and complicating threading model of a class, that is supposed to be "non thread safe". Why can not we keep this as non thread safe?

For example instead of setting this directly from the source thread, you could enqueue a mailbox action to unset it. Or even keeping the previous arrangement with acquiring manually `synchronized (lock)` in the source thread would be much better, as it wouldn't spill to the `OperatorChain`.

Reply:
I think that this would NOT introduce new "synchronisation" point because only visibility is an issue here (thread don't have to wait for each other).

I tried in `LegacySourceFunctionThread` 
```
mailboxProcessor
        .getMailboxExecutor(TaskMailbox.MAX_PRIORITY)
        .execute(
                () -> operatorChain.setIsStoppingBySyncSavepoint(false),
                "unsetIsStoppingBySyncSavepoint");
```
It failed ~33% runs because mailbox loop can be terminated earlier (`mailboxProcessor.allActionsCompleted()` is called concurrently).

Besides that, in my opinion it only moves the complexity to another place increasing it overall.

Using `synchronized` to prevent visibility issue seems somewhat non-obvious and error-prone to me.  But if you prefer I'll replace `volatile` with `synchronized`.

Reply:
It would be a "synchronisation" point, as that would be another place where one would have to start thinking about concurrency issues.

> Using synchronized to prevent visibility issue seems somewhat non-obvious and error-prone to me. But if you prefer I'll replace volatile with synchronized.

Yes, it's a hack, but at least you are not braking `@NonThreadSafe`tyness of the `OperatorChain`, and you do not spill this hack over the code base, as it's just limited to the `SourceStreamTask`.